### PR TITLE
test: add TAS e2e test infrastructure and basic tests

### PR DESCRIPTION
**Pull request by Ronkahn21**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR establishes the foundational infrastructure for Topology Aware Scheduling (TAS) e2e tests and includes two basic test scenarios.

**Infrastructure:**
- Adds 4-level topology hierarchy setup (zone → block → rack → host)
- Implements topology label application on k3d cluster nodes
- Adds KAI Topology verification utilities
- Adds ClusterTopology verification helpers
- Updates dependencies to KAI Scheduler v0.13.0-rc1
- Adds `topology-test` skaffold profile with TAS configuration
- Adds Makefile target for selective test execution (`TEST_PATTERN` support)

**Tests:**
- **TI1 (Topology Infrastructure)**: Verifies ClusterTopology and KAI Topology CRs are created correctly with proper 4-level hierarchy and node labels
- **BP1 (Basic Pattern)**: Tests multiple cliques with different topology constraints (rack-level and block-level packing)

This PR is part 1 of 4 in the TAS e2e test suite. Additional test scenarios will be added in follow-up PRs.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

**Dependencies:**
- This PR depends on PR #347 (code fixes for PCSG topology constraints) being merged first

**Test Verification:**
- All infrastructure files compile successfully
- Linter passes with 0 issues
- Test binary builds successfully with `-tags e2e`
- Two foundational tests included: TI1 (infrastructure) and BP1 (basic pattern)

**What's Next:**
- PR #349: Simple level & constraint tests (5 tests)
- PR #350: Scaling & hierarchical tests (5 tests)
- PR #351: Advanced scenarios & edge cases (4 tests)

**File Summary:**
- Infrastructure: 7 modified files (Makefile, skaffold, dependencies, setup files, workflow)
- Utilities: 3 new files (conversions, topology verification, KAI topology utilities)
- Tests: 1 new file with 2 tests + 1 helper function
- YAMLs: 1 test scenario file

#### Does this PR introduce a API change?

```release-note
NONE
```

#### Additional documentation e.g., enhancement proposals, usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: operator/e2e/setup/k8s_clusters.go [hunk 1]
```diff
@@ -1062,3 +1069,75 @@ func waitForWebhookReady(ctx context.Context, restConfig *rest.Config, logger *u
...
+	if idx <= 13 {
+		return "block-1"
+	}
+	return "block-2"
+}
+
+// getRackForNodeIndex returns the rack label for a given node index (0-based).
+// Distribution: 4 racks with 7 nodes each across 2 blocks.
+func getRackForNodeIndex(idx int) string {
+	rackRanges := []int{7, 13, 20, 27}
```
Comment (shayasoolin) [inline operator/e2e/setup/k8s_clusters.go]: These numbers here as well as the "13" in getBlockForNodeIndex make the code somewhat fragile, if we add more nodes in the future, even for non-TAS reasons.
I suggest to have one area with all the consts of number of nodes, number per block, etc. rather than hard-coding them in multiple places (again, including the place where we setup Skaffold total number of nodes).

File: operator/e2e/tests/setup.go [hunk 2]
```diff
@@ -181,12 +181,7 @@ func prepareTestCluster(ctx context.Context, t *testing.T, requiredWorkerNodes i
 			logger.Error("=== CLEANUP FAILURE - COLLECTING DIAGNOSTICS ===")
 			logger.Error("================================================================================")
 			CollectAllDiagnostics(diagnosticsTc)
-
-			// Mark cleanup as failed - this will cause all subsequent tests to fail immediately
```
Comment (shayasoolin) [inline operator/e2e/tests/setup.go:185]: Did you remove this intentionally? @gflarity just added it lately as a fix to the described issue.

File: operator/e2e/tests/topology_test.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+	workerCount := 0
+	for _, node := range nodes.Items {
+		if _, isControlPlane := node.Labels["node-role.kubernetes.io/control-plane"]; isControlPlane {
+			continue
+		}
+
+		workerCount++
+
+		// Verify zone label
```
Comment (shayasoolin) [inline operator/e2e/tests/topology_test.go]: Can we iterate over expectedKeys instead of having 4 similar if blocks?

File: operator/e2e/tests/setup.go [hunk 4]
```diff
@@ -642,11 +637,7 @@ func scalePCSGAcrossAllReplicas(tc TestContext, pcsName, pcsgName string, pcsRep
 
 // convertUnstructuredToTyped converts an unstructured map to a typed object
 func convertUnstructuredToTyped(u map[string]interface{}, typed interface{}) error {
```
Comment (shayasoolin) [inline operator/e2e/tests/setup.go]: This function became redundant, you can use the one from utils directly.

File: operator/e2e/tests/topology_test.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+// Scenario BP-1:
+// 1. Deploy workload with PCS (no constraint) containing 2 cliques:
+//   - worker-rack: packDomain=rack (3 pods)
+//   - worker-block: packDomain=block (4 pods)
+//
+// 2. Verify all 7 pods are scheduled successfully
+// 3. Verify worker-rack pods (3) are in the same rack
+// 4. Verify worker-block pods (4) are in the same block
+// 5. Verify different cliques can have independent topology constraints
+func Test_TAS_BP1_MultipleCliquesWithDifferentConstraints(t *testing.T) {
```
Comment (shayasoolin) [inline operator/e2e/tests/topology_test.go]: What does BP1 stands for? The convention in other tests is Test_TAS1_Multi... - no need for this extra short name with index. Just use TAS1, TAS2, ...

File: operator/e2e/tests/topology_test.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+	ctx := context.Background()
+
+	logger.Info("1. Initialize a 7-node Grove cluster for topology testing")
```
Comment (shayasoolin) [inline operator/e2e/tests/topology_test.go]: The cluster has 7 nodes and the workload has 7 pods. Not that it will necessarily help - but wouldn't it strengthen the test case if the cluster has more nodes than exactly what the workload expects? Since then there's more room for the scheduler to NOT pack the pods as the topology instructs, making the test is more reliable?

Comment (Ronkahn21) [reply operator/e2e/tests/topology_test.go]: you are correct, It was my mistake in the rest of the tests we use 28 nodes

File: operator/e2e/tests/setup.go (hunk 4 above)
Comment (Ronkahn21) [reply operator/e2e/tests/setup.go]: My mistake, I used it so I would not need to change the all the other tests. I will fix it

File: operator/e2e/tests/setup.go (hunk 2 above)
Comment (Ronkahn21) [reply operator/e2e/tests/setup.go:185]: Non intentionally, probably with rebase it cause this thanks for catching it.

File: operator/e2e/setup/k8s_clusters.go (hunk 1 above)
Comment (Ronkahn21) [reply operator/e2e/setup/k8s_clusters.go]: fixed

File: operator/e2e/tests/topology_test.go (hunk 3 above)
Comment (Ronkahn21) [reply operator/e2e/tests/topology_test.go]: fixed

File: operator/e2e/tests/topology_test.go (hunk 5 above)
Comment (Ronkahn21) [reply operator/e2e/tests/topology_test.go]: fixed

File: operator/e2e/dependencies.yaml [hunk 7]
```diff
@@ -18,25 +18,25 @@ images:
   
   # Kai Scheduler components
   - name: ghcr.io/nvidia/kai-scheduler/admission
-    version: v0.12.0
+    version: v0.13.0-rc1
```
Comment (gflarity) [inline operator/e2e/dependencies.yaml:21]: What's the ETA on 0.13?

File: operator/e2e/setup/skaffold.go [hunk 8]
```diff
@@ -315,3 +319,14 @@ func writeTemporaryKubeconfig(restConfig *rest.Config, logger *utils.Logger) (st
 	logger.Debugf("📄 Wrote temporary kubeconfig to: %s", tmpPath)
 	return tmpPath, cleanup, nil
 }
+
+// filterEnv filters out specified environment variables from the environment
+func filterEnv(env []string, keysToRemove ...string) []string {
```
Comment (gflarity) [inline operator/e2e/setup/skaffold.go:325]: Put some (Debug) logging for the things that get filtered?

File: operator/e2e/tests/topology_test.go [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+	corev1alpha1 "github.com/ai-dynamo/grove/operator/api/core/v1alpha1"
+	"github.com/ai-dynamo/grove/operator/e2e/setup"
+	"github.com/ai-dynamo/grove/operator/e2e/utils"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// deployWorkloadAndGetPods deploys workload, waits for pods to be ready, and returns the pod list
+func deployWorkloadAndGetPods(tc TestContext, expectedPods int) ([]v1.Pod, error) {
```
Comment (gflarity) [inline operator/e2e/tests/topology_test.go:34]: This might be handy for other tests to reuse in the future.

File: operator/e2e/setup/topology.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+	logger.Info("🏷️  Applying hierarchical topology labels to worker nodes...")
+
+	// Create clientset
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create clientset: %w", err)
+	}
+
+	// Get all worker nodes (filter by label set during cluster creation)
+	workerLabelSelector := fmt.Sprintf("%s=%s", WorkerNodeLabelKey, WorkerNodeLabelValue)
```
Comment (gflarity) [inline operator/e2e/setup/topology.go]: Nit, it's only a couple lines but I see them used a few times in the PR. Consider a small helper func.

File: operator/e2e/utils/topology.go [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+				pod.Name, topologyKey, actualValue, topologyKey, expectedValue)
+		}
+	}
+
+	logger.Infof("Verified %d pods are in same topology domain %s=%s", len(pods), topologyKey, expectedValue)
+	return nil
+}
+
+// WaitForPodsReady waits for the expected number of pods to be created and running
+func WaitForPodsReady(ctx context.Context, clientset kubernetes.Interface, namespace, labelSelector string, expectedPods int, timeout, interval time.Duration, logger *Logger) error {
```
Comment (gflarity) [inline operator/e2e/utils/topology.go]: Looks like this is duplicate of WaitForPods  in e2e/utils/k8s_client.go

**Comment by gflarity** [comment]
Just a few comments, the duplicate probably the most important to fix.  Just leaving comments to avoid blocking.

File: operator/e2e/dependencies.yaml (hunk 7 above)
Comment (Ronkahn21) [reply operator/e2e/dependencies.yaml:21]: I Asked the team. when the I get answer I will update here

File: operator/e2e/utils/topology.go (hunk 11 above)
Comment (Ronkahn21) [reply operator/e2e/utils/topology.go]: addressed

File: operator/e2e/setup/topology.go (hunk 10 above)
Comment (Ronkahn21) [reply operator/e2e/setup/topology.go]: fixed

File: operator/e2e/setup/topology.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+	// NodesPerZone is the number of nodes per zone.
+	NodesPerZone = 28
+	// NodesPerBlock is the number of nodes per block (28 / 2 blocks).
+	NodesPerBlock = 14
+	// NodesPerRack is the number of nodes per rack (28 / 4 racks).
+	NodesPerRack = 7
+)
+
+// GetZoneForNodeIndex returns the zone label for a given node index (0-based).
```
Comment (shayasoolin) [inline operator/e2e/setup/topology.go]: the "0-base" note here is misleading, as the zones are 1-based. maybe you meant that node index is 0-based.
in any case I suggest to have the zone index also 0-based, to avoid confusion.
same thing for block and others. 
did you have a specific reason to prefer 1-based?

File: operator/e2e/tests/topology_test.go [hunk 13]
```diff
@@ -24,19 +24,39 @@ import (
...
+	podList, err := listPods(tc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list pods: %w", err)
+	}
+
+	return podList.Items, nil
+}
+
+// Test_TAS1_TopologyInfrastructure verifies that the operator creates ClusterTopology and KAI Topology CRs at startup
 // Scenario TI-1 (Topology Infrastructure Setup):
```
Comment (shayasoolin) [inline operator/e2e/tests/topology_test.go]: given the removal of the extra test name, if you now don't use this scenario name "TI-1" anywhere else, I suggest to remove it from here as well.

File: operator/e2e/tests/topology_test.go [hunk 14]
```diff
@@ -112,8 +117,8 @@ func Test_TAS_TI1_TopologyInfrastructure(t *testing.T) {
 	logger.Info("🎉 Topology Infrastructure test completed successfully!")
 }
 
-// Test_TAS_BP1_MultipleCliquesWithDifferentConstraints tests PCS with multiple cliques having different topology constraints
-// Scenario BP-1:
+// Test_TAS2_MultipleCliquesWithDifferentConstraints tests PCS with multiple cliques having different topology constraints
+// Scenario TAS-2:
```
Comment (shayasoolin) [inline operator/e2e/tests/topology_test.go]: it's TAS2, but I think this line is not needed

File: operator/e2e/tests/topology_test.go [hunk 15]
```diff
@@ -210,14 +215,3 @@ func Test_TAS_BP1_MultipleCliquesWithDifferentConstraints(t *testing.T) {
 
 	logger.Info("🎉 BP-1: Multiple Cliques with Different Constraints test completed successfully!")
```
Comment (shayasoolin) [inline operator/e2e/tests/topology_test.go]: test name is no longer BP-1 (please go over other lines that refer to the old names and change them)

**Review by shayasoolin — APPROVED** [review]
